github/codeowners.py

```diff
@@ -81,73 +111,173 @@ def filter_codeowners_entries(
-                github_team_name=codeowner_entry[1:],
-                github_api=github_api,
-            )
+            yield Teamname(value=entry[1:])
+
+
+def resolve_email_addresses(
+    codeowners_entries: collections.abc.Collection[str],
+    github_api: GitHub | GitHubEnterprise,
+    prune_unique: bool=False,
+) -> typing.Generator[str, None, None]:
+    seen_email_addresses = []
+    for codeowner in iter_codeowners(codeowners_entries=codeowners_entries):
+        if isinstance(codeowner, EmailAddress):
+            mail = str(codeowner)
+
+            if prune_unique and mail in seen_email_addresses:
+                continue
+
+            yield mail
+            seen_email_addresses.append(mail)
+
+        elif isinstance(codeowner, Username):
+            if (mail := mail_address_for_user(
+                user_name=str(codeowner),
+                gh_api=github_api,
+            )):
+                if prune_unique and mail in seen_email_addresses:
+                    continue
+
+                yield mail
+                seen_email_addresses.append(mail)
+
+        elif isinstance(codeowner, Teamname):
+            org_name, team_slug = str(codeowner).split('/') # always of form 'org/name'
+            for member in resolve_team_members(
+                team_name=team_slug,
+                org=github_api.organization(org_name),
+                absent_ok=True,
+            ):
+                # team.members returns ShortUser
+                # email only contained in User
+                if (mail := mail_address_for_user(
+                    user_name=member.login,
+                    gh_api=github_api,
+                )):
+                    if prune_unique and mail in seen_email_addresses:
+                        continue
+
+                    yield mail
+                    seen_email_addresses.append(mail)
+
+
+def resolve_usernames(
+    codeowners_entries: collections.abc.Collection[str],
+    github_api: GitHub | GitHubEnterprise,
+    prune_unique: bool=False,
+) -> typing.Generator[str, None, None]:
+    seen_usernames = []
+    for codeowner in iter_codeowners(codeowners_entries=codeowners_entries):
+        if isinstance(codeowner, Username):
+            username = str(codeowner)
+            if prune_unique and username in seen_usernames:
+                continue
+
+            yield username
+            seen_usernames.append(username)
+
+        elif isinstance(codeowner, EmailAddress):
+            if (user := user_for_mail_address(
+                mail_address=str(codeowner),
+                gh_api=github_api,
+            )):
+                username = user.login
+                if prune_unique and username in seen_usernames:
+                    continue
+
+                yield username
+                seen_usernames.append(username)
+
+        elif isinstance(codeowner, Teamname):
+            org_name, team_slug = str(codeowner).split('/') # always of form 'org/name'
```

Comment:
as written above, mv to model-class; this is one of the few good object-oriented usages imo